armory/docker/images.py

```diff
@@ -32,3 +36,218 @@
+    if repo and repo != DOCKER_REPOSITORY:
+        return False
+    return name in IMAGE_MAP
+
+
+def get_armory_name(image_name: str):
+    """
+    Takes a docker image name, returns False if not a armory image
+        If an armory image, it will return `twosixarmory/<framework>:tag`
+    """
+    if not is_armory(image_name):
+        raise ValueError(f"Not an armory image name: {image_name}")
+    repo, name, tag = split_name(image_name)
+    repo = "twosixarmory"
+    if tag:  # tag is explicitly defined, use it
+        return join_name(repo, name, tag)
+    return IMAGE_MAP[name]
+
+
+def last_armory_release(image_name: str):
+    """
+    Return the image_name corresponding to the last armory major.minor.patch release
+        If the current image_name is a release, return the current
+    """
+    repo, name, tag = split_name(image_name)
+    if not repo or not tag:
+        raise ValueError("Must be a full repo/name:tag docker image_name")
+    tokens = tag.split(".")
+    if len(tokens) == 3:
+        return image_name
+    elif len(tokens) == 4:
+        # remove hash and decrement patch
+        major, minor, patch, _ = tokens
+        patch = int(patch)
+        if patch == 0:
+            raise ValueError(f"Tag {tag}: patch cannot be 0 for SCM with hash appended")
+        patch -= 1
+        patch = str(patch)
+        release_tag = ".".join([major, minor, patch])
+        return join_name(repo, name, release_tag)
+    else:
+        raise ValueError(
+            f"Tag {tag} must be in major.minor.patch[.hash] SCM version format"
+        )
+
+
+def ensure_image_present(image_name: str) -> str:
+    """
+    If image_name is available, return it. Otherwise, pull it from dockerhub.
+    """
+    log.trace(f"ensure_image_present {image_name}")
+    docker_client = docker.from_env()
+
+    if not is_armory(image_name):
+        log.trace(f"asking local docker for image {image_name}")
+        try:
+            docker_client.images.get(image_name)
+            log.success(f"found docker image {image_name}")
+            return image_name
+        except docker.errors.ImageNotFound:
+            log.trace(f"image {image_name} not found")
+        except requests.exceptions.HTTPError:
+            log.trace(f"http error when looking for image {image_name}")
+            raise
+
+        try:
+            pull_verbose(docker_client, image_name)
+            return image_name
+        except docker.errors.NotFound:
+            log.error(f"Image {image_name} could not be downloaded")
+            raise
+        except requests.exceptions.ConnectionError:
+            log.error("Docker connection refused. Is Docker Daemon running?")
+            raise
+
+    canon_image_name = get_armory_name(image_name)
+    log.info(f"Retrieved canonical image name for {image_name} as {canon_image_name}")
+
+    prev_release = last_armory_release(canon_image_name)
+    if canon_image_name != prev_release:  # currently on hashed dev branch
```

Comment:
This is spooky mysterious action as is currently in discussion. The saving grace here is that the decision tree progress is logged.